### PR TITLE
Resource group refactoring

### PR DESCRIPTION
**Pull request by cberner**
refactoring to support making resource group management pluggable.

### CODE REVIEW

**Comment by cberner** [comment]
Looks good as long as the tests pass

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupId.java [hunk 1]
```diff
@@ -13,37 +13,46 @@
...
+        this(singletonList(requireNonNull(name, "name is null")));
     }
 
     public ResourceGroupId(ResourceGroupId parent, String name)
     {
-        this(ImmutableList.<String>builder().addAll(requireNonNull(parent, "parent is null").segments).add(requireNonNull(name, "name is null")).build());
+        this(concatenate(requireNonNull(parent, "parent is null").segments, requireNonNull(name, "name is null")));
+    }
+
+    private static List<String> concatenate(List<String> list, String element)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupId.java]: Maybe call it "append"?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroup.java [hunk 2]
```diff
@@ -25,37 +25,68 @@
...
 
+    /**
+     * Threshold on total distributed CPU usage after which max running queries will be reduced.
+     */
     void setSoftCpuLimit(Duration limit);
 
     Duration getHardCpuLimit();
 
+    /**
+     * Threshold on total distributed CPU usage after which new queries
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroup.java]: It's a bit confusing that the queueing is controlled by the "hard" cpu limit vs the "soft" memory limit

**Comment by martint** [comment]
looks good

**Pull request merged**